File: logger/src/impl/java/io/appmetrica/analytics/logger/impl/WordBreakFinder.java
package io.appmetrica.analytics.logger.impl;

import androidx.annotation.NonNull;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

class WordBreakFinder {

    private final Pattern goodLineEndPattern = Pattern.compile("[\\p{Space},;]");

    public int find(@NonNull String input, int startOffset, int endOffset) {
        Matcher matcher = goodLineEndPattern.matcher(input);
        return findLastIndex(matcher, startOffset, endOffset);
    }

    private int findLastIndex(@NonNull Matcher matcher, int start, int end) {
        int result = -1;

        if (end >= start) {
            int baseLine = (end - start) / 2 + start;
            matcher.region(baseLine, end);
            if (matcher.find()) {
                result = findLastIndex(matcher, end);
            } else {
                matcher.region(start, baseLine);
                if (matcher.find()) {
                    result = findLastIndex(matcher, baseLine);
                }
            }
        }

        return result;
    }

    private int findLastIndex(@NonNull Matcher matcher, int end) {
        int firstFound = matcher.start();
        int closerFound = findLastIndex(matcher, firstFound + 1, end);
        return closerFound == -1 ? firstFound : closerFound;
    }
}

File: common-logger/src/main/java/io/appmetrica/analytics/logger/common/impl/WordBreakFinder.kt
package io.appmetrica.analytics.logger.common.impl

import java.util.regex.Matcher
import java.util.regex.Pattern

internal class WordBreakFinder {

    private val goodLineEndPattern: Pattern by lazy {
        Pattern.compile("[\\p{Space},;]")
    }

    fun find(input: String, startOffset: Int, endOffset: Int): Int {
        return findLastIndex(goodLineEndPattern.matcher(input), startOffset, endOffset)
    }

    private fun findLastIndex(matcher: Matcher, start: Int, end: Int): Int {
        if (end < start) {
            return -1
        }
        val baseLine = (end - start) / 2 + start
        matcher.region(baseLine, end)
        return if (matcher.find()) {
            findLastIndex(matcher, end)
        } else {
            matcher.region(start, baseLine)
            if (matcher.find()) {
                findLastIndex(matcher, baseLine)
            } else {
                -1
            }
        }
    }

    private fun findLastIndex(matcher: Matcher, end: Int): Int {
        val firstFound = matcher.start()
        val closerFound = findLastIndex(matcher, firstFound + 1, end)
        return if (closerFound == -1) firstFound else closerFound
    }
}
